src/shell_utils.cpp

```diff
@@ -0,0 +1,169 @@
+        std::cerr << "Failed to execute command." << std::endl;
+        exit(EXIT_FAILURE);	// In case execvp fails
+    }
+}
+
+// Executes a command with the given command path, arguments, and background execution flag
+void executeCommand(const std::string& commandPath, const std::vector<std::string>& arguments, bool runInBackground) {
+    pid_t pid = fork();
+    if (pid < 0) {
+        std::cerr << "Fork failed." << std::endl;
+        return;
+    }
+    if (pid == 0) {
+        // Child process
+        executeChildProcess(commandPath, arguments);
+    }
+    else {
+        // Parent process
+        if (!runInBackground) {
+            int status;
+            waitpid(pid, &status, 0); // Wait for the child process to complete
+            // Update the job status based on the process exit status
+            bool isRunning = !WIFEXITED(status) || (WEXITSTATUS(status) == 0);
+            updateJobStatus(pid, isRunning);
+        }
+        else {
+            // Add the child process details (pid, command, etc.) to a list of background jobs
+            Job job(commandPath, pid, true);
+            backgroundJobs.push_back(job);
+        }
+    }
+}
+
+// Update the job status based on the process ID
+void updateJobStatus(int pid, bool isRunning) {
+    for (Job& job : backgroundJobs) {
+        if (job.pid == pid) {
+            job.isRunning = isRunning;
+            break;
+        }
+    }
+}
+
+// Prints the details of a job (command, PID, status)
+void printCommandDetails(const Job& job) {
+    std::cout << "Command: " << job.command << std::endl;
+    std::cout << "PID: " << job.pid << std::endl;
+    std::cout << "Status: " << (job.isRunning ? "Running" : "Finished") << std::endl;
+    std::cout << std::endl;
+}
+
+// Executes the 'myjobs' command to display information about background jobs
+void executeMyJobsCommand() {
+    if (backgroundJobs.empty()) {
+        std::cout << "No background jobs." << std::endl;
+        return;
+    }
+
+    // Iterate over the background jobs and print their details
+    for (const Job& job : backgroundJobs) {
+        printCommandDetails(job);
+    }
+}
+
+// Executes the 'cd' command to change the current working directory
+void executeCdCommand(const std::vector<std::string>& tokens) {
+    if (tokens.size() > 1) {
+        if (chdir(tokens[1].c_str()) != 0) {
+            std::cerr << "Failed to change directory." << std::endl;
+        }
+    }
+    else {
+        std::cerr << "Missing argument for cd command." << std::endl;
+    }
+}
+
+// Searches the PATH environment variable for the given command and returns the full path if found
+std::string searchPathForCommand(const std::string& commandPath, const char* pathEnv) {
+    std::string pathString(pathEnv);
+    std::vector<std::string> paths = split(pathString, ':');
+    for (const std::string& path : paths) {
+        std::string fullPath = path + '/' + commandPath;
+        if (access(fullPath.c_str(), X_OK) == 0) {
```

Comment:
maybe use this https://en.cppreference.com/w/cpp/filesystem/exists or even one of the "is_file" methods